File: scripts/import-manulex.mjs
// Convertit un export brut Manulex (colonnes : lemme, niveau_1, ... niveau_6,
// une valeur de fréquence par colonne — format distribué par les auteurs
// pour usage académique) en src/data/frequenceLexicale.json, au format
// attendu par src/lib/manulex.js : { motNormalise: 'P1' | 'P2' | ... }
//
// Usage : node scripts/import-manulex.mjs chemin/vers/export-manulex.csv
//
// Le fichier source n'est pas fourni avec ce dépôt : il doit être obtenu
// séparément (voir scripts/README.md) sous licence académique auprès du
// projet Manulex.

import { mkdirSync, readFileSync, writeFileSync } from 'node:fs'

const NIVEAUX_COLONNES = ['P1', 'P2', 'P3', 'P4', 'P5', 'P6']

function normaliser(mot) {
  return mot
    .toLowerCase()
    .normalize('NFD').replace(/[̀-ͯ]/g, '')
    .replace(/[^a-z-]/g, '')
}

// Nombre de colonnes attendu : lemme + une colonne de fréquence par niveau.
const NB_COLONNES_ATTENDU = 1 + NIVEAUX_COLONNES.length

function convertir(cheminCsv) {
  const contenu = readFileSync(cheminCsv, 'utf-8').replace(/\r/g, '')
  const lignes = contenu.split('\n').filter(Boolean)
  const [enTete, ...donnees] = lignes

  // Le format exact du fichier source (délimiteur, colonnes) dépend de ce
  // que l'export académique fournit réellement — non vérifiable tant
  // qu'aucun fichier réel n'a été obtenu. Un en-tête au mauvais nombre de
  // colonnes est le signe le plus probable d'un mauvais délimiteur (ex.
  // point-virgule au lieu de virgule) : on arrête plutôt que de produire un
  // dataset silencieusement corrompu.
  const colonnesEnTete = enTete.split(',')
  if (colonnesEnTete.length !== NB_COLONNES_ATTENDU) {
    console.error(
      `En-tête inattendu : ${colonnesEnTete.length} colonne(s) trouvée(s), ${NB_COLONNES_ATTENDU} attendue(s) ` +
      `(lemme + ${NIVEAUX_COLONNES.join(', ')}). Le fichier utilise peut-être un autre délimiteur que la virgule.`
    )
    process.exit(1)
  }

  const resultat = {}
  let lignesIgnorees = 0
  for (const ligne of donnees) {
    const colonnes = ligne.split(',')
    if (colonnes.length !== NB_COLONNES_ATTENDU) {
      lignesIgnorees++
      continue
    }

    const lemme = normaliser(colonnes[0])
    if (!lemme) {
      lignesIgnorees++
      continue
    }

    for (let i = 0; i < NIVEAUX_COLONNES.length; i++) {
      const frequence = Number(colonnes[i + 1])
      if (Number.isFinite(frequence) && frequence > 0) {
        resultat[lemme] = NIVEAUX_COLONNES[i]
        break
      }
    }
  }
  return { resultat, lignesIgnorees }
}

const [, , cheminCsv] = process.argv
if (!cheminCsv) {
  console.error('Usage : node scripts/import-manulex.mjs chemin/vers/export-manulex.csv')
  process.exit(1)
}

let dataset, lignesIgnorees
try {
  ;({ resultat: dataset, lignesIgnorees } = convertir(cheminCsv))
} catch (err) {
  console.error(`Impossible de lire ${cheminCsv} : ${err.message}`)
  process.exit(1)
}

mkdirSync('src/data', { recursive: true })
writeFileSync('src/data/frequenceLexicale.json', JSON.stringify(dataset, null, 2))
console.log(`${Object.keys(dataset).length} mots exportés vers src/data/frequenceLexicale.json`)
if (lignesIgnorees > 0) {
  console.log(`${lignesIgnorees} ligne(s) ignorée(s) (colonnes manquantes ou lemme vide).`)
}

File: scripts/import-manulex.py
#!/usr/bin/env python3
"""
Convertit le fichier Excel officiel Manulex (Lété, Sprenger-Charolles & Colé,
2004) en src/data/frequenceLexicale.json, au format attendu par
src/lib/manulex.js : { motNormalise: 'P1' | 'P2' | 'P3' }.

Source : https://www.manulex.org/fr/downloads.html — "Fichier Excel de Manulex"
(licence Creative Commons BY-NC-SA 3.0 — non-commercial, paternité obligatoire,
partage à l'identique). Citer Lété, B., Sprenger-Charolles, L. & Colé, P. (2004)
partout où ces données sont utilisées.

Format réel de la feuille LEMMES (vérifié le 2026-09-14, 23 813 lignes) :
colonnes LEMMES, NLET, SYNT, puis pour chaque tranche CP / CE1 / CE2-CM2 :
F (fréquence brute), D (dispersion), U (fréquence/million), SFI. Il n'y a PAS
de colonne par année scolaire individuelle : CE2, CM1 et CM2 sont regroupés
en une seule tranche "CE2-CM2".

Mapping vers les niveaux FWB :
  CP       -> P1  (CP = système français, équivalent 1re primaire en FWB)
  CE1      -> P2
  CE2-CM2  -> P3  (première année de la tranche, faute de détail par année ;
                    un mot qui n'apparaît qu'en CM2 sera donc traité comme
                    "connu dès P3", légèrement optimiste pour un enseignant
                    de P3 strict)

LIMITE IMPORTANTE : Manulex ne couvre que les manuels de lecture du primaire
français (CP à CM2, donc P1 à P3-P5 FWB). Il n'y a AUCUNE donnée secondaire
(S1-S6) dans cette base — la détection de mots hors-niveau restera sans effet
pour les textes destinés au secondaire, quel que soit l'import réalisé ici.

Un mot n'est retenu à un niveau que s'il y apparaît au moins SEUIL_FREQUENCE
fois dans l'échantillon Manulex de cette tranche, pour écarter le bruit des
occurrences isolées (nom de personnage dans un seul manuel, coquille...).

Usage : python scripts/import-manulex.py chemin/vers/Manulex.xls
Nécessite le paquet xlrd (pip install xlrd) — pas de nouvelle dépendance npm :
le fichier xls de Manulex est au format binaire historique (BIFF/OLE2), et le
paquet npm 'xlsx' (SheetJS) a des failles connues sans correctif disponible
(prototype pollution, ReDoS) — écarté pour cette raison.
"""

import json
import re
import sys
import unicodedata
from pathlib import Path

SEUIL_FREQUENCE = 3

NIVEAUX_COLONNES = [
    ('P1', 'CP F'),
    ('P2', 'CE1 F'),
    ('P3', 'CE2-CM2 F'),
]


def normaliser(mot):
    mot = mot.lower()
    mot = ''.join(c for c in unicodedata.normalize('NFD', mot) if unicodedata.category(c) != 'Mn')
    return re.sub(r'[^a-z-]', '', mot)


def convertir(chemin_xls):
    import xlrd

    wb = xlrd.open_workbook(chemin_xls)
    sh = wb.sheet_by_name('LEMMES')
    en_tete = sh.row_values(0)

    try:
        idx_lemme = en_tete.index('LEMMES')
        idx_colonnes = [(niveau, en_tete.index(col)) for niveau, col in NIVEAUX_COLONNES]
    except ValueError as exc:
        raise SystemExit(
            f"Colonne attendue absente de l'en-tête ({exc}). "
            f"En-tête trouvé : {en_tete}\n"
            "Le format du fichier Manulex a peut-être changé depuis l'écriture de ce script."
        )

    resultat = {}
    lignes_ignorees = 0
    for r in range(1, sh.nrows):
        row = sh.row_values(r)
        lemme = normaliser(str(row[idx_lemme]))
        if not lemme:
            lignes_ignorees += 1
            continue

        trouve = False
        for niveau, idx in idx_colonnes:
            frequence = row[idx]
            if isinstance(frequence, (int, float)) and frequence >= SEUIL_FREQUENCE:
                resultat.setdefault(lemme, niveau)
                trouve = True
                break
        if not trouve:
            lignes_ignorees += 1

    return resultat, lignes_ignorees


def main():
    if len(sys.argv) != 2:
        print('Usage : python scripts/import-manulex.py chemin/vers/Manulex.xls', file=sys.stderr)
        sys.exit(1)

    chemin_xls = sys.argv[1]
    resultat, lignes_ignorees = convertir(chemin_xls)

    sortie = Path(__file__).resolve().parent.parent / 'src' / 'data' / 'frequenceLexicale.json'
    sortie.parent.mkdir(parents=True, exist_ok=True)
    sortie.write_text(json.dumps(resultat, ensure_ascii=False, indent=2, sort_keys=True), encoding='utf-8')

    print(f"{len(resultat)} mots exportés vers {sortie}")
    print(f"{lignes_ignorees} ligne(s) ignorée(s) (lemme vide ou fréquence < {SEUIL_FREQUENCE} partout).")


if __name__ == '__main__':
    main()
